gson/src/main/java/com/google/gson/JsonDeserializer.java

```diff
@@ -51,7 +51,9 @@
  * class IdDeserializer implements JsonDeserializer&lt;Id&gt;() {
  *   public Id deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
  *       throws JsonParseException {
- *     return new Id((Class)typeOfT, id.getValue());
+ *     JsonObject jsonObject = json.getAsJsonObject();
+ *     long idValue = jsonObject.get("value").getAsLong();
+ *     return new Id((Class) typeOfT, idValue);
```

Comment:
Or actually maybe this logic should be:
```suggestion
 *     long idValue = json.getAsJsonPrimitive().getAsLong();
 *     return new Id((Class) typeOfT, idValue);
```
This would match the logic in [`JsonSerializer`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/JsonSerializer.html), and also what the sentences a few line above this code block in the documentation say.

Reply:
we cannot use getAsJsonPrimitive on a Json Object. 
In the example mentioned above, json is ID object and will read as JsonObject and can get variables from that object.

Refer the screenshoot: 
<img width="533" alt="Screen Shot 2022-09-09 at 10 58 33 AM" src="https://user-images.githubusercontent.com/55409379/189384741-eee85e1f-3a90-4384-bd1b-922830fdf91b.png">

Reply:
Yes that is how the `Id` class would have been serialized without custom serializer, however the sample serializer in the documentation of `JsonSerializer` serializes it as `20`.

And the `JsonDeserializer` documentation a few lines above the code sample you have edited says (emphasis mine):
> The default deserialization of `Id(com.foo.MyObject.class, 20L)` will require the JSON string to be `{"clazz":"com.foo.MyObject","value":20}`. Suppose, you already know the type of the field that the `Id` will be deserialized into, and hence just **want to deserialize it from a JSON string `20`**. You can achieve that by writing a custom deserializer:

Reply:
updaed

Reply:
Thanks! Personally I would have added a separate `long idValue` variable to increase readability, but it does not matter that much. If you like you can change it though.